File: Algorithms/Examples/Part3/Java/Stack.java
package Java;

public class Stack {

	private StackNode head;

	class StackNode {
		int data;
		StackNode next;

		StackNode(int val) {
			data = val;
			next = null;
		}
	}

	public Stack() {
		head = null;
	}

	public boolean isEmpty() {
		if (head == null)
			return true;
		else
			return false;
	}

	public void push(int x) {
		StackNode newNode = new StackNode(x);

		if (head == null) {
			head = newNode;
		} else {
			StackNode temp = head;
			head = newNode;
			newNode.next = temp;
		}
		assert(!isEmpty());
		assert(top() != -1);
	}

	public int pop() {
		if (!isEmpty()) {
			int popped = head.data;
			head = head.next;
			return popped;
		} else {
			return -1;
		}
	}

	public int top() {
		if (!isEmpty())
			return head.data;
		else
			return -1;
	}

	public static void main(String[] args) {
		Stack stk = new Stack();
		for (int i = 0; i < 10; i++)
			stk.push(i);
		for (int i = 0; i < 11; i++)
			System.out.println(stk.pop() + ", " + stk.top());
		System.out.println("Empty? " + stk.isEmpty());
	}

}

File: Algorithms/Examples/Part3/C/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

Stack* createStack(unsigned maxsize) {
	Stack* stk = malloc(sizeof(Stack));
	stk->index = -1;
	stk->maxsize = maxsize;
	stk->arr = (int*) malloc(stk->maxsize * sizeof(int));
	return stk;
}

int isempty(Stack* stk) {
	return (stk->index == -1);
}

void push(int x, Stack* stk) {
	if (stk->index == stk->maxsize - 1) {
		return;
	} else {
		stk->arr[++stk->index] = x;
	}
}

int pop(Stack* stk) {
	if (isempty(stk))
		return -1;
	else
		return stk->arr[stk->index--];
}

int top(Stack* stk) {
	if (isempty(stk))
		return -1;
	else
		return stk->arr[stk->index];
}

